src/audio/mixer.c

```diff
@@ -255,11 +257,13 @@ static int mixer_trigger(struct comp_dev *dev, int cmd)
 			/* Mixer and everything downstream is active */
 			dev->state = COMP_STATE_ACTIVE;
 			break;
+#if CONFIG_IPC_MAJOR_3
 		case COMP_TRIGGER_PRE_START:
 			/* Mixer and downstream components might or might not be active */
 			if (mixer_source_status_count(dev, COMP_STATE_ACTIVE) ||
 			    mixer_source_status_count(dev, COMP_STATE_PAUSED))
 				return PPL_STATUS_PATH_STOP;
+#endif
```

Comment:
@lyakh  why add these code ? I don't understand

Reply:
because both mixer and mux/demux connect multiple pipelines. So they have a more complex state machine than "normal" components with just one sink and one source. Every component is checking when getting a trigger to transition to a different state, whether that transition is allowed. And usually allowed transitions are mostly "sequential" and the same for all components: from initialised to active and back. But with mixer and (de)mux it can happen that, e.g. the component is already in an active state and it's receiving a prepare or start trigger on another pipeline. So their state machine has to handle such triggers too as valid and act accordingly.
After I've added `PRE_START` and `PRE_RELEASE` commands mixer and (de)mux now have to handle two additional commands, otherwise they'd report invalid command for their current state.